numba/tests/test_np_functions.py

```diff
@@ -4893,6 +4893,7 @@ def create_harcoded_variant(self, basefunc, ty):
         eval(compile(funcstr, '<string>', 'exec'))
         return locals()['foo']
 
+    @unittest.skipIf(numpy_version >= (1, 24), "NumPy < 1.24 required")
```

Comment:
This also fails for me on numpy 1.23.4 (openSUSE Tumbleweed packaging):

```
[ 2303s] ======================================================================
[ 2303s] ERROR: test_MachAr (numba.tests.test_np_functions.TestNPMachineParameters)
[ 2303s] ----------------------------------------------------------------------
[ 2303s] Traceback (most recent call last):
[ 2303s]   File "/usr/lib/python3.8/site-packages/numba/tests/test_np_functions.py", line 4783, in test_MachAr
[ 2303s]     self.check(machar, attrs)
[ 2303s]   File "/usr/lib/python3.8/site-packages/numba/tests/test_np_functions.py", line 4762, in check
[ 2303s]     got = cfunc(*args)
[ 2303s]   File "/usr/lib/python3.8/site-packages/numba/core/dispatcher.py", line 468, in _compile_for_args
[ 2303s]     error_rewrite(e, 'typing')
[ 2303s]   File "/usr/lib/python3.8/site-packages/numba/core/dispatcher.py", line 409, in error_rewrite
[ 2303s]     raise e.with_traceback(None)
[ 2303s] numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
[ 2303s] Use of unsupported NumPy function 'numpy.MachAr' or unsupported use of the function.
[ 2303s] 
[ 2303s] File "../../../../../../usr/lib/python3.8/site-packages/numba/tests/test_np_functions.py", line 102:
[ 2303s] def machar(*args):
[ 2303s]     return np.MachAr()
[ 2303s]     ^
[ 2303s] 
[ 2303s] During: typing of get attribute at /usr/lib/python3.8/site-packages/numba/tests/test_np_functions.py (102)
[ 2303s] 
[ 2303s] File "../../../../../../usr/lib/python3.8/site-packages/numba/tests/test_np_functions.py", line 102:
[ 2303s] def machar(*args):
[ 2303s]     return np.MachAr()
[ 2303s]     ^
[ 2303s] 
[ 2303s] 
[ 2303s] ======================================================================
[ 2303s] FAIL: test_np_MachAr_deprecation_np122 (numba.tests.test_np_functions.TestNPMachineParameters)
[ 2303s] ----------------------------------------------------------------------
[ 2303s] Traceback (most recent call last):
[ 2303s]   File "/usr/lib/python3.8/site-packages/numba/tests/support.py", line 599, in inner
[ 2303s]     self.subprocess_test_runner(test_module=self.__module__,
[ 2303s]   File "/usr/lib/python3.8/site-packages/numba/tests/support.py", line 584, in subprocess_test_runner
[ 2303s]     self.assertEqual(status.returncode, 0, streams)
[ 2303s] AssertionError: 1 != 0 : 
[ 2303s] captured stdout: 
[ 2303s] captured stderr: E
[ 2303s] ======================================================================
[ 2303s] ERROR: test_np_MachAr_deprecation_np122 (numba.tests.test_np_functions.TestNPMachineParameters)
[ 2303s] ----------------------------------------------------------------------
[ 2303s] Traceback (most recent call last):
[ 2303s]   File "/usr/lib/python3.8/site-packages/numba/tests/support.py", line 607, in inner
[ 2303s]     func(self)
[ 2303s]   File "/usr/lib/python3.8/site-packages/numba/tests/test_np_functions.py", line 4836, in test_np_MachAr_deprecation_np122
[ 2303s]     f()
[ 2303s]   File "/usr/lib/python3.8/site-packages/numba/core/dispatcher.py", line 468, in _compile_for_args
[ 2303s]     error_rewrite(e, 'typing')
[ 2303s]   File "/usr/lib/python3.8/site-packages/numba/core/dispatcher.py", line 409, in error_rewrite
[ 2303s]     raise e.with_traceback(None)
[ 2303s] numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
[ 2303s] Use of unsupported NumPy function 'numpy.MachAr' or unsupported use of the function.
[ 2303s] 
[ 2303s] File "../../../../../../usr/lib/python3.8/site-packages/numba/tests/test_np_functions.py", line 4835:
[ 2303s]     def test_np_MachAr_deprecation_np122(self):
[ 2303s]         <source elided>
[ 2303s]                                     category=NumbaDeprecationWarning,)
[ 2303s]             f = njit(lambda : np.MachAr().eps)
[ 2303s]             ^
[ 2303s] 
[ 2303s] During: typing of get attribute at /usr/lib/python3.8/site-packages/numba/tests/test_np_functions.py (4835)
[ 2303s] 
[ 2303s] File "../../../../../../usr/lib/python3.8/site-packages/numba/tests/test_np_functions.py", line 4835:
[ 2303s]     def test_np_MachAr_deprecation_np122(self):
[ 2303s]         <source elided>
[ 2303s]                                     category=NumbaDeprecationWarning,)
[ 2303s]             f = njit(lambda : np.MachAr().eps)
[ 2303s]             ^
[ 2303s] 
[ 2303s] 
[ 2303s] ----------------------------------------------------------------------
[ 2303s] Ran 1 test in 0.271s
[ 2303s] 
[ 2303s] FAILED (errors=1)
[ 2303s] 
[ 2303s] 
[ 2303s] ----------------------------------------------------------------------
[ 2303s] Ran 9684 tests in 2243.604s
```

Reply:
I could have messed something up for old NumPy versions because I've only been testing with 1.24 in the matrix here - I've opened https://github.com/numba/numba/pull/8690 to run CI on this branch but with the other NumPy versions to check for this and maybe any other problems.

Reply:
@bnavigator thanks for the report, this seems a bit strange as NumPy 1.23 definitely has `MachAr` still. Maybe there's an issue with Numba's overload construction for it that is triggered in the openSUSE builds. Perhaps open a new issue to discuss/track? Thanks!

Reply:
> I could have messed something up for old NumPy versions because I've only been testing with 1.24 in the matrix here - I've opened #8690 to run CI on this branch but with the other NumPy versions to check for this and maybe any other problems.

I think there's potentially something else going on. Looking at the NumPy source, there's `MachAr` exported in both 1.23 and 1.24.

Reply:
There's no `np.MachAr` in 1.24, but `np.core.MachAr` is there - it's been left in but is not public / documented, AIUI. I think that will be removed too in a future release.

Reply:
Ouch. Looks like I have some outdated patch which still includes

```diff
@@ -4203,7 +4207,7 @@ def _gen_np_machar():
         return impl
 
 
-_gen_np_machar()
+# _gen_np_machar()
 
 
 def generate_xinfo(np_func, container, attr):
```

Reply:
@gmarkall Thanks, I was indeed looking under `numpy.core`, where it still exists.

@bnavigator Ah, thanks for checking, guess that's the root cause of the problem!

Reply:
Sorry for the noise. My mistake. I can confirm that the current PR with 7d3826e as the latest commit works for both NumPy 1.23.4 and after an update to 1.24.1 when packaging numba 0.56.4 for openSUSE Tumbleweed.

Reply:
@bnavigator no worries, thanks for checking that it's now all working! This is essentially an independent verification of the update in this patch, which is really useful to have!